app/controllers/carto/oauth_provider_controller.rb

```
@@ -42,11 +43,11 @@ class OauthProviderController < ApplicationController
     rescue_from OauthProvider::Errors::BaseError, with: :rescue_oauth_errors
 
     def consent
+      raise OauthProvider::Errors::LoginRequired.new if @login_required_error
       return create_authorization_code if @oauth_app_user.try(:authorized?, @scopes)
+      raise OauthProvider::Errors::AccessDenied.new if silent_flow?
 
-      if @oauth_app_user
-        return create_authorization_code if @oauth_app_user.authorized?(@scopes)
-      elsif @oauth_app
+      if @oauth_app && !@oauth_app_user
```

This one is on me, but I'm pretty sure that we can remove this check. If `oauth_app` is not present, the code should raise an exception in one of the `before_action`s and never reach here.

---

I was talking about `@oauth_app`. i.e: I think this only needs to be `unless @oauth_app_user` :)